File: functions/src/domain/member/repositories/factory/MemberRepositoryFactoryImpl.ts
import { RepositoryFactory } from '../../../shared/RepositoryFactory';
import { MemberRepositoryFactoryContext } from './MemberRepositoryFactoryContext';
import { MemberRepository } from '../member.repository';
import { MemberRepositoryImpl } from '../../../../infrastructure/repositories/member.repository.impl';
import { FirestoreMemberDataSource } from '../../../../infrastructure/datasources/firestore/member.datasource';
import { firestore } from 'firebase-admin';

/**
 * Factory for creating member repositories
 */
export class MemberRepositoryFactoryImpl implements RepositoryFactory<MemberRepository, MemberRepositoryFactoryContext> {
  /**
   * Creates a member repository instance
   * @param context - Context containing projectId for member repository creation
   * @returns MemberRepository instance
   */
  create(context?: MemberRepositoryFactoryContext): MemberRepository {
    if (!context || !context.projectId) {
      throw new Error('MemberRepositoryFactoryContext with projectId is required');
    }

    const db = firestore();
    const dataSource = new FirestoreMemberDataSource(db);
    return new MemberRepositoryImpl(dataSource, context.projectId);
  }
}

File: data/data_core/src/main/java/com/example/data_core/repository/factory/MemberRepositoryFactoryImpl.kt
package com.example.data_core.repository.factory

import com.example.data_core.datasource.remote.MemberRemoteDataSource
import com.example.data_core.repository.base.MemberRepositoryImpl
import com.example.domain.repository.RepositoryFactory
import com.example.domain.repository.base.MemberRepository
import com.example.domain.repository.factory.context.MemberRepositoryFactoryContext
import javax.inject.Inject

class MemberRepositoryFactoryImpl @Inject constructor(
    private val memberRemoteDataSource: MemberRemoteDataSource
) : RepositoryFactory<MemberRepositoryFactoryContext, MemberRepository> {

    override fun create(input: MemberRepositoryFactoryContext): MemberRepository {
        return MemberRepositoryImpl(
            memberRemoteDataSource = memberRemoteDataSource,
            factoryContext = input,
        )
    }
}
